[LegacySVG] clip-path inset is incorrectly applied to the svg element.

smfr opened the pull request:
#### 891e5f19413790c00aea80246816bc341cf78f3f
<pre>
[LegacySVG] clip-path inset is incorrectly applied to the svg element.
<a href="https://bugs.webkit.org/show_bug.cgi?id=227704">https://bugs.webkit.org/show_bug.cgi?id=227704</a>
<a href="https://rdar.apple.com/80516912">rdar://80516912</a>

Reviewed by Said Abou-Hallawa.

`clip-path()` on the SVG root element on the legacy code path (LBSE disabled) must be applied
in RenderLayer, because the clipping needs to clip the CSS-rendered borders and shadows on the
element, and the RenderLayer code path gets the geometry right.

We then have to not clip in `SVGRenderingContext::prepareToRenderSVGContent()` when applied
to the LegacyRenderSVGRoot.

In addition, fix RenderView&apos;s `rendererObscuresBackground()` to check for the legacy renderer
type, otherwise we&apos;d end up with unfilled areas.

I verified that this behaves correctly with CSS and property-applied clip-path on the SVG root
inside the SVG, and when applied via CSS to SVG embedded in HTML.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/clip-path/clip-path-marginBox-1d.html:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::setupClipPath):
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::rendererObscuresBackground):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::clipContextToCSSClippingArea):
* Source/WebCore/rendering/svg/SVGRenderingContext.cpp:
(WebCore::SVGRenderingContext::prepareToRenderSVGContent):

Canonical link: <a href="https://commits.webkit.org/281893@main">https://commits.webkit.org/281893@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/64d8aa2775c8765c6b3df33c8205e1979fd863ab

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61301 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40662 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13882 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65251 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11850 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63431 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48340 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12125 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49529 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8230 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/63632 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37802 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53101 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30365 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34475 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10328 "Found 1 new test failure: imported/w3c/web-platform-tests/content-security-policy/reporting-api/reporting-api-report-only-sends-reports-on-violation.https.sub.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10763 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56291 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10624 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66982 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5248 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10418 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56903 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5271 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53067 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57109 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13677 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4332 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36466 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37549 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38643 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37293 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/rendering/RenderView.cpp:
```diff
@@ -375,7 +375,7 @@ static inline bool rendererObscuresBackground(const RenderElement& rootElement)
         return false;
 
-    if (rootElement.hasClipPath() && rootElement.isRenderSVGRoot())
+    if (rootElement.hasClipPath() && (rootElement.isRenderSVGRoot() || rootElement.isLegacyRenderSVGRoot()))
```
`isRenderOrLegacyRenderSVGRoot()`?

shallawa commented on Source/WebCore/rendering/svg/SVGRenderingContext.cpp:
```diff
@@ -152,7 +152,7 @@ void SVGRenderingContext::prepareToRenderSVGContent(RenderElement& renderer, Pai
 
     auto* clipper = resources->clipper();
-    if (!hasCSSClipping && clipper) {
+    if (!hasCSSClipping && clipper && !is<LegacyRenderSVGRoot>(renderer)) {
```
I think for the `<svg>` root element, `hasCSSClipping` has to be true. So maybe we can `ASSERT(!is<LegacyRenderSVGRoot>)` inside the if-statement instead.

smfr committed:
[LegacySVG] clip-path inset is incorrectly applied to the svg element.
https://bugs.webkit.org/show_bug.cgi?id=227704
rdar://80516912

Reviewed by Said Abou-Hallawa.

`clip-path()` on the SVG root element on the legacy code path (LBSE disabled) must be applied
in RenderLayer, because the clipping needs to clip the CSS-rendered borders and shadows on the
element, and the RenderLayer code path gets the geometry right.

We then have to not clip in `SVGRenderingContext::prepareToRenderSVGContent()` when applied
to the LegacyRenderSVGRoot.

In addition, fix RenderView's `rendererObscuresBackground()` to check for the legacy renderer
type, otherwise we'd end up with unfilled areas.

I verified that this behaves correctly with CSS and property-applied clip-path on the SVG root
inside the SVG, and when applied via CSS to SVG embedded in HTML.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/clip-path/clip-path-marginBox-1d.html:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::setupClipPath):
* Source/WebCore/rendering/RenderView.cpp:
(WebCore::rendererObscuresBackground):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::clipContextToCSSClippingArea):
* Source/WebCore/rendering/svg/SVGRenderingContext.cpp:
(WebCore::SVGRenderingContext::prepareToRenderSVGContent):

Canonical link: https://commits.webkit.org/281893@main
